designs/component-config.md

```diff
@@ -0,0 +1,312 @@
+- allow using and upgrading older configurations avoiding breaking changes in flags
+
+### Links to Open Issues
+
+- [#518 Provide a ComponentConfig to tweak the Manager](https://github.com/kubernetes-sigs/controller-runtime/issues/518)
+- [#207 Reduce command line flag boilerplate](https://github.com/kubernetes-sigs/controller-runtime/issues/207)
+- [#722 Implement ComponentConfig by default & stop using (most) flags](https://github.com/kubernetes-sigs/kubebuilder/issues/722)
+
+### Goals
+
+- Provide an interface for pulling configuration data out of exposed `ComponentConfig` types (see below for implementation)
+- Provide a new `ctrl.NewFromComponentConfig()` function for initializing a manager
+- Provide an embeddable `RuntimeConfig` type for easily authoring `ComponentConfig` types
+- Provide an `DefaultControllerConfig` to make the switch easier
+
+### Non-Goals/Future Work
+
+- `kubebuilder` implementation and design in another PR
+- Changing the default `controller-runtime` implementation
+- Dynamically reloading `ComponentConfig` object
+- Providing `flags` interface and overrides
+
+## Proposal
+
+The `ctrl.Manager` _SHOULD_ support loading configurations from `ComponentConfig` like objects. 
+An interface for that object with getters for the specific configuration parameters is created to bridge existing patterns.
+
+Without breaking the current `ctrl.NewManager` which uses an exported `ctrl.Options{}` the `manager.go` can expose a new func, `NewFromComponentConfig()` this would be able to loop through the getters to hydrate an internal `ctrl.Options{}` and pass that into `New()`.
+
+```golang
+//pkg/manager/manager.go
+
+// ManagerConfiguration defines what the ComponentConfig object for ControllerRuntime needs to support
+type ManagerConfiguration interface {
+	GetSyncPeriod() *time.Duration
+
+	GetLeaderElection() bool
+	GetLeaderElectionNamespace() string
+	GetLeaderElectionID() string
+
+	GetLeaseDuration() *time.Duration
+	GetRenewDeadline() *time.Duration
+	GetRetryPeriod() *time.Duration
+
+	GetNamespace() string
+	GetMetricsBindAddress() string
+	GetHealthProbeBindAddress() string
+
+	GetReadinessEndpointName() string
+	GetLivenessEndpointName() string
+
+	GetPort() int
+	GetHost() string
+
+	GetCertDir() string
+}
+
+func NewFromComponentConfig(config *rest.Config, scheme *runtime.Scheme, filename string, managerconfig ManagerConfiguration) (Manager, error) {
+	codecs := serializer.NewCodecFactory(scheme)
+    if err := decodeComponentConfigFileInto(codecs, filename, componentconfig); err != nil {
+		
+	}
+	options := Options{}
+
+	if scheme != nil {
+		options.Scheme = scheme
+	}
+
+	// Loop through getters
+	if managerconfig.GetLeaderElection() {
+			managerconfig.GetLeaderElection()
+	}
+	// ...
+
+	return New(config, options)
+}
+```
+
+#### ComponentConfig Load Order
+
+![ComponentConfig Load Order](/designs/images/component-config-load.png)
+
+#### Embeddable ComponentConfig Type
```

Comment:
I think that makes sense. I'm hoping to pull as much of the work from https://github.com/kubernetes/component-base/blob/master/config/v1alpha1/types.go for example pulling the `LeaderElectionConfiguration` potentially this is a better place to push more of those configurations. 

Fair point that `meta` might not be the right place long term for it, have any suggestions? maybe `config` -> `pkg/api/config/v1alpha1/types.go` like I had used for the `DefaultControllerConfiguration`…

UX wise I would expect that someone could embed this into their type as a natural way to extend to contain non-manager specific components. Eg:

```
package config

import (
  "time"

  metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
  metav1alpha1 "sigs.k8s.io/controller-runtime/pkg/apis/meta/v1alpha1"
) 

type ControllerNameConfigurationSpec struct {
  metav1alpha1.RuntimeConfig `json:",inline"`

  // ... Other configurations that are specific to the custom controller
}

type ControllerNameConfiguration struct {
  metav1.TypeMeta

  Spec ControllerNameConfigurationSpec `json:"spec"`
}
```

Reply:
Should we call these `pkg/api/config/v1alpha1/types.go` then?

Reply:
This doesn't belong in the meta api.

Reply:
Why put this in k/k at all? Is there a reason we can't offer these types, which are for controller-runtime users, from the controller-runtime repo?

Wrt. component-base, I'd like to take a conservative approach to what we put in there. The purpose of that repo is to consolidate foundational code that was spread out across k/k, and improve some of those frameworks within the confines of the core project. Since the controllers in k/k (to my knowledge...which isn't comprehensive) aren't (yet) implemented on controller-runtime, I'm not sure it makes sense to put these types in component-base (yet).

Reply:
sorry that import is from `sigs.k8s.io/controller-runtime/pkg/apis/meta/v1alpha1` in `contoller-runtime` I was not expecting that to end up in k/k.

Reply:
Ah, makes sense :)
Thanks for the update.

Reply:
From `controller-runtime` user perspective having a meta package in our repository makes sense to offer basic types that are configurable when users implement ComponentConfig in their managers, +1